### PR TITLE
topology : avoid indexer hash possible collision

### PR DESCRIPTION
**Pull request by nplanel**
Signed-off-by: Nicolas PLANEL <nplanel@redhat.com>

### CODE REVIEW

File: topology/graph/indexer.go [hunk 1]
```diff
@@ -34,11 +34,6 @@ type NodeHasher func(n *Node) map[string]interface{}
 
 // Hash computes the hash of the passed parameters
 func Hash(values ...interface{}) string {
-	if len(values) == 1 {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply topology/graph/indexer.go]: 1/ hash is called with
Hash("storage", "1") = 99a1014ec1e468221e92afe2c06a72b5

2/ what happen if Hash is called with a field name StorageKey (outside of skydive scope)
Hash("99a1014ec1e468221e92afe2c06a72b5") = 99a1014ec1e468221e92afe2c06a72b5

As actual code there are very poor chance of collision but the actual code introduce one

[the remark this answers is not in the document]
Comment (lebauce) [reply topology/graph/indexer.go]: You are not supposed to call Hash with a different number of arguments. This function is not a universal hash method but a method dedicated to the indexer.

[the remark this answers is not in the document]
Comment (nplanel) [reply topology/graph/indexer.go]: yeah maybe ... but one day someone would like to use the package graph hash will do  graph.Hash() and will not expect a 1 argument string trick

[the remark this answers is not in the document]
Comment (lebauce) [reply topology/graph/indexer.go]: Not sure it's useful to fix problem that may (or may not) occur in the future. If so then your PR is the wrong fix : it just slows down everything instead. A fix should be to make Hash a method of Indexer or be renamed.

[the remark this answers is not in the document]
Comment (nplanel) [reply topology/graph/indexer.go]: I agree the case is very limited today but the hash is exported, on our case (one string value) speed vs collision free... don't forgot that reflection is more likely kill the pref more than the hash algorithm. 
Your point on performance is more likely than structhash will build a new string from a string (adding quote and version) and using md5

Yes would be better to add a callback hash function like hashNodes is pass trough

... so I propose
 1/ use murmurhash algorithm
 2/ push a PR on structhash to support murmurhash (as they stringify values before md5/sha1 it
 3/ having an indexer hash modifier like indexer.SetHash() or via NewIndexer
 
In the second step as the arguments values types are fixed, a specific hash function per indexer type could be added to limit the reflection (need some testing on this point)

**Comment by safchain** [comment]
closing this for now until finding a better fix

**Pull request closed without merging**